### ISSUE TITLE
Popover Feature Improvements

### ISSUE THREAD

**Issue by colinmorelli** [ng-bootstrap/ng-bootstrap#837]
The popover implementation is currently missing bits that were present in Angular 1's implementation that are a huge help when using them:

- **Detect when the popover is open** Could use either the "two way" `[(isOpen)]` syntax, or just exposing event outputs for `(open)` and `(close)`. Either way, some method of notifying the parent component when popovers are toggled would be helpful.
- **Close other popovers when one opens** This was default behavior in angular 1 IIRC. I'm not sure if it should be default, or an option, but my assumption is that most implementations would prefer having only one open at a time.
- **Close on outside click**. The combination of this missing, plus the previous point, means that components that want to control multiple popovers such that only one is open at any time get fairly ugly.
- **Close on escape** I personally don't care much about this, but I figured I'd mention it for completeness. For many users, the default behavior to close something that they don't want open is to use the escape key. Popovers probably would benefit from responding to this much like modals do.

I believe a couple of these are fairly low hanging fruit, and I'd be happy to put up a PR for them (particularly I'm referring to items 1, 3, and possibly 4 - if it's something you're interested in adding.)

I believe the second item, though, is going to require a coordinating type of service much like modals have, to determine which are open/closed.

**Comment by pkozlowski-opensource** [comment]
> Detect when the popover is open Could use either the "two way" [(isOpen)] syntax, or just exposing event outputs for (open) and (close). Either way, some method of notifying the parent component when popovers are toggled would be helpful.

Would you mind sharing a real-life use-case for this? I'm not saying we shouldn't do it, I'm just saying that we need to balance larger API surface with the frequency / importance of the need.

The other request sound reasonable. PRs would be awesome but what would be even better before would be:
* a separate issue for each point (this way it is easier to chop off chunks of work and execute it in parallel)
* a short design (could be just an entry in the issue) with the proposed API and impl considerations. This way we could validate the approach and avoid wasted effort on both sides (coding and code reviews).

If you care about those things by all means step in!

**Comment by colinmorelli** [comment]
Hey @pkozlowski-opensource thanks for the quick response.

Use case for #1 is simple: I want to add a CSS class to the button that triggers the popover to change the style of the button when the popover is open. The most "flexible" way to accomplish that IMO is simply to provide events for when it's opened/closed.

**Comment by pkozlowski-opensource** [comment]
> Use case for #1 is simple: I want to add a CSS class to the button that triggers the popover to change the style of the button when the popover is open. The most "flexible" way to accomplish that IMO is simply to provide events for when it's opened/closed.

Sounds reasonable. But here is my alternative proposal that would satisfy this use-case as well - we could add and expose a public `isOpen()` method on the popover instance so you could do: `<div ngbPopover="..." #myPop="ngbPopover">` and then in your button: `<button [class.foo]="myPop.isOpen()>`. How does it sound?

> Happy to make separate issues for the various points here

Would appreciate. This will be much easier to manage.

> As for the design proposal: are you looking for that with each item or are there some in particular you're more concerned with?

No, no particular concerns on this point, just would like to make sure that we agree on the API and have a chance to discuss any potential impl pitfalls.

**Comment by colinmorelli** [comment]
Let's continue discussion for this in #841 (will follow with other issues soon)
